sdk/monitor/monitor-ingestion/src/logsIngestionClient.ts

```diff
@@ -0,0 +1,146 @@
+      options?.userAgentOptions && options?.userAgentOptions.userAgentPrefix
+        ? `${options?.userAgentOptions.userAgentPrefix} ${packageDetails}`
+        : `${packageDetails}`;
+    this._dataClient = new GeneratedDataCollectionClient(tokenCredential, this.endpoint, {
+      ...options,
+      credentialScopes: credentialOptions?.credentialScopes ?? defaultIngestionScope,
+      userAgentOptions: {
+        userAgentPrefix,
+      },
+    });
+    // adding gziping policy bcz this is a single method client which needs gzipping
+    this._dataClient.pipeline.addPolicy(GZippingPolicy);
+  }
+
+  /**
+   * See error response code and error response message for more detail.
+   * @param ruleId - The immutable Id of the Data Collection Rule resource.
+   * @param streamName - The streamDeclaration name as defined in the Data Collection Rule.
+   * @param logs - An array of objects matching the schema defined by the provided stream.
+   * @param options - The options parameters.
+   */
+  async upload(
+    ruleId: string,
+    streamName: string,
+    logs: Record<string, any>[],
+    options?: UploadOptions
+  ): Promise<UploadResult> {
+    // TODO: Do we need to worry about memory issues when loading data for 100GB ?? JS max allocation is 1 or 2GB
+
+    // This splits logs into 1MB chunks
+    const chunkArray: any[] = this.splitDataToChunks(logs);
+    const noOfChunks = chunkArray.length;
+    let concurrency = 1;
+    if (options?.maxConcurrency && options?.maxConcurrency > 1) {
+      concurrency = options?.maxConcurrency;
+    }
+
+    const uploadResult: UploadResult = {
+      errors: [],
+      uploadStatus: "Success",
+    };
+
+    const errorsArray: Record<string, any>[] = [];
+    const uploadCallback = async (x: any): Promise<void> => {
+      try {
+        await this._dataClient.upload(ruleId, streamName, x, {
+          contentEncoding: "gzip",
+        });
+      } catch (e: any) {
+        errorsArray.push({ error: e, log: x });
+      }
+    };
+
+    for await (const _i of asyncPool(concurrency, chunkArray, uploadCallback)) {
+      /* eslint-disable no-empty */
+    }
+    for (const errorsObj of errorsArray) {
+      uploadResult.errors.push({
+        responseError: errorsObj.error,
+        failedLogs: errorsObj.log,
+      });
+    }
+    if (uploadResult.errors.length === 0) {
+      return uploadResult;
+    } else if (uploadResult.errors.length < noOfChunks && uploadResult.errors.length > 0) {
+      uploadResult.uploadStatus = "PartialFailure";
+      return uploadResult;
+    } else {
+      throw new FailedLogsIngestionError("All logs failed for ingestion", uploadResult.errors);
+    }
+  }
+
+  /**
+   * @internal
+   */
+  splitDataToChunks(logs: Record<string, any>[]): any[] {
+    let chunk: any[] = [];
+    const chunkArray: any[] = [];
+    let size = 0;
+    const maxBytes = 1000000;
+    for (const element of logs) {
+      const elementSize = JSON.stringify(element).length * 4;
```

Comment:
this isn't necessarily true for paired unicode characters, right?

Reply:
do you know if there's a way to detect and compute the size of the paired unicode characters in JS?

Reply:
They say there is no way to detect a unicode character - https://stackoverflow.com/questions/25292855/counting-unicode-characters-in-javascript Do they have different size in the memory? @xirzec

Reply:
We changed the way we calculate the size now.